### PR TITLE
Adding git-secrets-scan-action

### PR DESCRIPTION
**Pull request by ojasjoshi**
*Issue #, if available:*

*Description of changes:*
This simple action uses [git secrets](https://github.com/awslabs/git-secrets) to scan your repository for secrets being commited to public.

*Tests*
Tested on personal forks with/without committing fake aws_credentials.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: git-secrets-scan-action/LICENSE.md
Comment (mm318) [inline git-secrets-scan-action/LICENSE.md]: This should be Apache 2.0 and attributed to Amazon (if applicable to license text).

File: git-secrets-scan-action/Dockerfile
Comment (mm318) [inline git-secrets-scan-action/Dockerfile]: Is AWS CLI actually required?

File: git-secrets-scan-action/entrypoint.sh
Comment (mm318) [inline git-secrets-scan-action/entrypoint.sh]: Let's do `--scan-history` as well.

File: git-secrets-scan-action/README.md
Comment (mm318) [inline git-secrets-scan-action/README.md]: I think we should check if the directory pointed to by this environment variable is correct. Currently, if this environment variable is incorrect, the action will pass by default, right? But we want it to fail in that case.

Comment (mm318) [reply git-secrets-scan-action/README.md]: A correct directory should be at least a non-empty path string that points to an existent path.

File: git-secrets-scan-action/entrypoint.sh
Comment (mm318) [inline git-secrets-scan-action/entrypoint.sh]: As mentioned in https://github.com/aws-robotics/aws-robomaker-github-actions/pull/17#discussion_r502231832, we should add a check here for the value of `${GITHUB_WORKSPACE}`.

File: git-secrets-scan-action/entrypoint.sh
Comment (ojasjoshi) [reply git-secrets-scan-action/entrypoint.sh]: `--scan-history` actually takes very long time, especially if the commit history is too large. I am not sure if that is required as part of the CI workflow.. we can do a manual `--scan-history` once before merging this change. All the changes following this should be checked by the CI

File: git-secrets-scan-action/Dockerfile
Comment (ojasjoshi) [reply git-secrets-scan-action/Dockerfile]: Removed

File: git-secrets-scan-action/LICENSE.md
Comment (ojasjoshi) [reply git-secrets-scan-action/LICENSE.md]: Done.

File: git-secrets-scan-action/README.md
Comment (ojasjoshi) [reply git-secrets-scan-action/README.md]: I believe GH actions guarantees that this env variable is set to the workspace directory.. but a good catch to point that the default behavior here (when the path is empty/incorrect) would be that the action passes with exit code 0.

Comment (ojasjoshi) [reply git-secrets-scan-action/README.md]: Added a check to see if GITHUB_WORKSPACE is set  and points to a valid directory..also checks that the repo is checked-out using actions/checkout.. Tested that the action now fails in both cases..

File: git-secrets-scan-action/LICENSE.md
Comment (mm318) [reply git-secrets-scan-action/LICENSE.md]: By the way, the license for `s3-cp-action` doesn't need to change because that's effectively a fork of https://github.com/jakejarvis/s3-sync-action.

File: git-secrets-scan-action/entrypoint.sh
Comment (mm318) [reply git-secrets-scan-action/entrypoint.sh]: Oh, interesting.

I think it is a legitimate concern though, because if we push multiple commits and some intermediate commit has credentials, then `git-secrets` won't catch it but it will still be accessible by someone who's detail oriented.

Does `--scan-history` take longer than build and bundle? What if we made `git-secrets` run in parallel to our CI, instead of as a serial step in our CI?

Comment (ojasjoshi) [reply git-secrets-scan-action/entrypoint.sh]: Oh yep, that is definitely a concern, good point. 

I tested `--scan-history` on our SAs(which is primarily our use case) and it doesn't take a lot of time(< 30s). I believe for our potential future use-cases, the time should be similar.  In that case, let us keep `--scan-history` in the action and run it sequentially in our workflow. In case we see issues with significantly larger CI times in the future we can find a workaround at that time.

File: git-secrets-scan-action/README.md
Comment (mm318) [inline git-secrets-scan-action/README.md]: This GitHub Action doesn't actually use AWS CLI.

File: git-secrets-scan-action/README.md
Comment (mm318) [reply git-secrets-scan-action/README.md]: Also, `GITHUB_WORKSPACE` can be set explicitly for this step in the workflow, right? I ask because sometimes `actions/checkout` may be used multiple times (so does `GITHUB_WORKSPACE` end up taking on the value of the path that we want) and/or not used to check out the sources that we want to actually scan (for example, the way `ros-tooling/action-ros-ci` is used).

File: git-secrets-scan-action/entrypoint.sh
Comment (mm318) [inline git-secrets-scan-action/entrypoint.sh]: I don't think the directory we're checking has to come from `actions/checkout`.

File: git-secrets-scan-action/README.md
Comment (ojasjoshi) [reply git-secrets-scan-action/README.md]: oh right, missed to remove that

File: git-secrets-scan-action/entrypoint.sh
Comment (ojasjoshi) [reply git-secrets-scan-action/entrypoint.sh:19]: Per https://github.com/actions/checkout#Checkout-multiple-repos-side-by-side, any checkout action occurs in `GITHUB_WORKSPACE` by default. (or `GITHUB_WORKSPACE/path` if a path is provided in actions/checkout).

Comment (ojasjoshi) [reply git-secrets-scan-action/entrypoint.sh]: Added `path` input variable to accommodate the repository being checked out at a different location within `GITHUB_WORKSPACE`

File: git-secrets-scan-action/README.md
Comment (ojasjoshi) [reply git-secrets-scan-action/README.md]: `GITHUB_WORKSPACE` can be explicitly set although the user might not know the path to the workspace inside the docker container which GH actions set by default.

I believe that actions/checkout is probably the universal way to checkout your repo inside GH actions. Per https://github.com/actions/checkout#Checkout-multiple-repos-side-by-side, any checkout action occurs within `GITHUB_WORKSPACE` by default. (or `GITHUB_WORKSPACE/path` if a path is provided in actions/checkout). 

```
# Relative path under $GITHUB_WORKSPACE to place the repository
    path: ''
```
But now it looks like we are not accommodating for that `path` variable.

Comment (ojasjoshi) [reply git-secrets-scan-action/README.md]: Added `path` input variable to accommodate the repository being checked out at a different location within `GITHUB_WORKSPACE`

Comment (ojasjoshi) [reply git-secrets-scan-action/README.md:48]: While testing out the action I figured that, 
In actions/checkout@v1, `path` is the absolute path where the repo will get checked-out.
In actions/checkout@v2, `path` is relative to `GITHUB_WORKSPACE` where the repo will get checked-out.

So now I guess the best way for us to deal with this is use absolute path to the repo. A issue with that being the user might not know the path to the workspace inside the docker container which GH actions set by default( via `GITHUB_WORKSPACE`)

Comment (ojasjoshi) [reply git-secrets-scan-action/README.md:48]: While testing out the action I figured that, 
In actions/checkout@v1, `path` is the absolute path where the repo will get checked-out.
In actions/checkout@v2, `path` is relative to `GITHUB_WORKSPACE` where the repo will get checked-out.

A way out of this is to check for both, local path and absolute path. If neither point to a valid directory, we exit with failure.

Comment (ojasjoshi) [reply git-secrets-scan-action/README.md:48]: While testing out the action I figured that, 
In actions/checkout@v1, input variable `path` is not used. 
In actions/checkout@v2, input variable `path` is relative to `GITHUB_WORKSPACE` where the repo will get checked-out.

A potential way out of this is to check for both, local path and absolute path. If neither point to a valid directory, we exit with failure. 

We can always ask the user for the absolute path to the repo, however, if the repo is checked-out wrt `GITHUB_WORKSPACE`, the user might not know the absolute path.

Comment (ojasjoshi) [reply git-secrets-scan-action/README.md]: While testing out the action I figured that, 
In actions/checkout@v1, input variable `path` is not used. 
In actions/checkout@v2, input variable `path` is relative to `GITHUB_WORKSPACE` where the repo will get checked-out.

A potential way out of this is to check for the path per `actions/checkout@v1` or per `actions/checkout@v2` based on if `path` is set or not(see README). If the path does not point to a valid directory, exit with failure. This however, would *probably* not work in the case of `actions/checkout@v2` being used for nested repositories.

As a simple alternative, we can always ask the user for the absolute path to the repo, however, if the repo is checked-out wrt `GITHUB_WORKSPACE`, the user might not know the absolute path.

edit: testing on actions/checkout@v2.. also resolving [https://github.com/actions/checkout/issues/126]

File: git-secrets-scan-action/action.yml
Comment (mm318) [inline git-secrets-scan-action/action.yml]: I think the `path` input should be just be relative to the current working directory (I think that's how all of our actions have been behaving).

The current working directory is independent of the `$GITHUB_WORKSPACE` environment variable, right? And I think neither `actions/checkout@v1` or `actions/checkout@v2` modifies the current working directory. That would also avoid dealing with the behavioral difference between `actions/checkout@v1` and `actions/checkout@v2`.

File: git-secrets-scan-action/README.md
Comment (mm318) [reply git-secrets-scan-action/README.md]: I made a relevant comment at https://github.com/aws-robotics/aws-robomaker-github-actions/pull/17#discussion_r504952382.

File: git-secrets-scan-action/action.yml
Comment (mm318) [inline git-secrets-scan-action/action.yml]: This should say "Relative path under the current working directory or an absolute path to the repository that should be scanned".

**Review by mm318 — APPROVED** [review]
Approving in advance, since https://github.com/aws-robotics/aws-robomaker-github-actions/pull/17#discussion_r505097152 should be easy to address.

**Pull request merged**
